### DISCUSSION TITLE
I converted [0, 255] ranged pixel values to [0, 1] range with this command.

### DISCUSSION THREAD

**Discussion by bemoregt** [ImageMagick/ImageMagick#2381] [Help]
Hi, All.

I converted [0, 255] ranged pixel values to [0, 1] range with this command.

for fn in *.jpg; do convert -threshold 128 -evaluate divide 255 "${fn}" "${fn}"; done

But, I met the pixel color [0, 2] range. 

What's wrong to me?
and How can I get [0, 1] range?

Thanks
Best,
@bemoregt.

**Comment by snibgo** [comment]
What version of IM, on what platform? What is the text output of "convert -version"?

Are you sure your pixels are in the range [0,255]?

When IM reads a file, it stores the image in memory where pixel values are usually in the range 0 to QuantumRange, which is 2^q-1 where q is the Q-number of your IM build. For Q8 QuantumRange is 255, for Q16 QuantumRange is 65535, etc.

So if you really want pixels values in the range 0 to 1, divide by QuantumRange.

```
convert in.png -evaluate Divide %[fx:QuantumRange] out.png
```

**Reply by bemoregt** [reply]
It also failed.

the result image have just 0 values.

--;

**Comment by bemoregt** [comment]
What version of IM, on what platform? 
>> My IM version is
ImageMagick 7.0.10-25 Q16 x86_64 2020-08-01
>> and Mac OSX Mojave 10.14.6

Are you sure your pixels are in the range [0,255]?
>> Sure.

Thanks. I'll try that.

Best,
@bemoregt.

**Comment by snibgo** [comment]
For IM v7, I suggest you use "magick", not "convert".

I repeat:
> What is the text output of "convert -version"? Or "magick -version"?

Perhaps you don't have HDRI, in which case values will be integers, so you will have no values between 0.0 and 1.0.

What is your exact command? Are you saving to a floating-point output? If not, then again you will only get integer values.

**Reply by bemoregt** [reply]
What is your exact command?
>> for fn in *.jpg; do convert -threshold 128 -evaluate divide 255 "${fn}" "${fn}"; done

**Comment by snibgo** [comment]
Another consideration: if your IM is Q16, then QuantumLevel is 65535. So dividing by QuantumLevel gives values between 0 and 1, out of 65535. If you save this to an 8-bit format such as JPG, IM will divide all values by 257. So the maximum value sent to the JPG encoder will be 1/257 (on a scale of 0 to 255). So all the values will be zero.

But you still haven't shown the text output of "-version", so I can only guess.

It might help if you explained what you are trying to do.

**Reply by bemoregt** [reply]
Hi, @snibgo 

What is the text output of "convert -version"? Or "magick -version"?
>> convert -version
Version: ImageMagick 7.0.10-23 Q16 x86_64 2020-07-05 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(3.1) 
Delegates (built-in): bzlib freetype gslib heic jng jp2 jpeg lcms lqr ltdl lzma openexr png ps tiff webp xml zlib

>> magick -version
Version: ImageMagick 7.0.10-23 Q16 x86_64 2020-07-05 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(3.1) 
Delegates (built-in): bzlib freetype gslib heic jng jp2 jpeg lcms lqr ltdl lzma openexr png ps tiff webp xml zlib

Thanks.
Best,
@bemoregt.

**Comment by snibgo** [comment]
Okay, so you have IM Q16, with HDRI.

When IM reads a JPEG, or any image, it multiplies values so that white is at QuantumRange, which is 65535 for Q16. You threshold at 128, which is a small value compared to 65535, so we expect most (or all) pixel values will become white (65535). Then you divide by 255, so most pixel values will become 65535/255 = 257.

When IM writes a Q16 image to an 8-bit file such as JPG, it first divides values by 257. So most or all of your pixels will become 257/257 = 1 out of 256.

Instead of thresholding at 128, perhaps you want to threshold at 50% of QuantumRange, ie 50% of white. If so, use "-threshold 50%".

**Reply by bemoregt** [reply]
It's also same result.

It makes [0,2] levels also.

**Comment by snibgo** [comment]
JPEG is a lossy format. This means pixels may be changed when they are written to a JPEG file. So your pixel values may be in [0,1] before the write, but may have a wider range when written to JPEG.
